doc/developer/design/20240508_subsource_schemas_after_planning.md

```diff
@@ -0,0 +1,113 @@
+    -   https://github.com/MaterializeInc/materialize/pull/26881
+    -   https://github.com/MaterializeInc/materialize/issues/26999
+
+## The Problems
+
+-   To complete #20208, we have to know the schema of the Kafka topic. This
+    schema is most readily knowable after planning the `SOURCE`.
+-   Subsource planning (generating `CreateSubsourceStatement`) currently happens
+    during purification before planning the `SOURCE`.
+-   Purification's current structure also means that we have to implement some
+    bits in purification (for new sources) and some in sequencing (when
+    reconciling additional subsource's definitions with the current source).
+    This would be cleaner if we described the semantics of e.g. source options
+    somewhere like `SourceConnection`.
+
+## Success Criteria
+
+-   We generate `CreateSubourceStatement`s _after_ planning the
+    `CreateSourceStatement`. This unblocks the ability to theoretically create
+    subsources for Kafka's output.
+-   We demonstrate, using PostgreSQL sources as an example, that it is possible
+    to separate "planning" and purification's async concerns, e.g.
+    `PurificationError` wraps some other error that expresses the semantics of
+    handling PostgreSQL `TEXT COLUMNS`.
+
+## Out of Scope
+
+-   Actually creating subsources for Kafka sources.
+-   Schema migrations
+-   Fully separating concerns w/r/t purification, i.e. I am not planning to
+    address [MySQL source's blending of
+    concerns](https://github.com/MaterializeInc/materialize/issues/26999),
+    though I think that the spirit of this change should include that task.
+-   Supporting `ALTER SOURCE...ADD SUBSOURCE` for MySQL, though that should be
+    relatively simple to do.
+
+## Solution Proposal
+
+1.  Modularize source planning. Source planning will become more complex and the
+    current code structure stretches many people's ability to comprehend what's
+    happening.
+2.  Move the API for planning that occurs during purification into a spot more
+    semantically appropriate, e.g. let the `SouceConnection` trait define how
+    its options should be allowed to be used. e.g. we should provide common
+    errors that both `PurificationError` (for new sources) and `AdapterError`
+    (for additional subsources) can wrap to signal that there are issues with
+    source definitions.
+3.  Restructure purification such that we track the names of the sources we've
+    been requested to create, as well as the "upstream" objects to which they
+    refer––i.e. instead of generating `Vec<CreateSubsourceStatement>` we should
+    generate `BTreeMap<SubsourceName, UpstreamObjectName>`.
+4.  Add a fn to the `SourceConnection` trait like:
+
+    ```rust
+    fn get_output_schema_for_name(
+        &self,
+        upstream_object_name: UpstreamObjectName
+    ) -> Result<RelationDesc, Err>
+    ```
+
+    This will let us ask for the given schema of a subsource after the source
+    itself is planned.
```

Comment:
How do we validate the options like TEXT COLUMNS on the top-level source plan before we've requested the schema from the relevant subsources? Or will we start allowing invalid references in those options in the top-level source statement and throw an error when we plan all the subsources?

Reply:
`TEXT COLUMNS` and `IGNORE COLUMNS` only need to validate that they refer to valid columns in the source's "subsource catalog." That should just be done as part of normalizing them––i.e. when we figure out which object they refer to, we should write down that fully qualified name, and if we can't find it, we should error.

Reply:
idk if this makes things any clearer, but I [added a section](https://github.com/MaterializeInc/materialize/pull/26881/files#diff-7271ca5f7922adef08c51331d47c305a1e8fb810e2273a8d4b2b80668bf8e3aeR104) around some of the unification I'm hoping to achieve between creating and modifying sources. This might just end up being a conceptual, rather than concrete, framework for me. idk how gnarly the typing will get.